hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClientHandler.java

```diff
@@ -0,0 +1,123 @@
+package org.apache.hadoop.fs.azurebfs.services;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import org.apache.hadoop.fs.azurebfs.AbfsConfiguration;
+import org.apache.hadoop.fs.azurebfs.constants.AbfsServiceType;
+import org.apache.hadoop.fs.azurebfs.contracts.exceptions.InvalidUriException;
+import org.apache.hadoop.fs.azurebfs.extensions.EncryptionContextProvider;
+import org.apache.hadoop.fs.azurebfs.extensions.SASTokenProvider;
+import org.apache.hadoop.fs.azurebfs.oauth2.AccessTokenProvider;
+import org.apache.hadoop.util.Preconditions;
+
+import static org.apache.hadoop.fs.azurebfs.constants.FileSystemUriSchemes.ABFS_BLOB_DOMAIN_NAME;
+import static org.apache.hadoop.fs.azurebfs.constants.FileSystemUriSchemes.ABFS_DFS_DOMAIN_NAME;
+import static org.apache.hadoop.fs.azurebfs.utils.UriUtils.changeUrlFromBlobToDfs;
+import static org.apache.hadoop.fs.azurebfs.utils.UriUtils.changeUrlFromDfsToBlob;
+
+/**
+ * AbfsClientHandler is a class that provides a way to get the AbfsClient
+ * based on the service type.
+ */
+public class AbfsClientHandler {
+
+  private AbfsServiceType defaultServiceType;
+  private AbfsServiceType ingressServiceType;
+  private final AbfsDfsClient dfsAbfsClient;
+  private final AbfsBlobClient blobAbfsClient;
+
+  public AbfsClientHandler(final URL baseUrl,
```

Comment:
Can be merged into a single constructor as we are anyway passing null values, check can be made inside consructor. Just a suggestion :-

public AbfsClientHandler(final URL baseUrl,
                         final SharedKeyCredentials sharedKeyCredentials,
                         final AbfsConfiguration abfsConfiguration,
                         final AccessTokenProvider tokenProvider,
                         final SASTokenProvider sasTokenProvider,
                         final EncryptionContextProvider encryptionContextProvider,
                         final AbfsClientContext abfsClientContext) throws IOException {
    if (baseUrl == null || sharedKeyCredentials == null || abfsConfiguration == null ||
            encryptionContextProvider == null || abfsClientContext == null) {
        throw new IllegalArgumentException("BaseUrl, SharedKeyCredentials, AbfsConfiguration, EncryptionContextProvider, and AbfsClientContext must be non-null");
    }

    this.dfsAbfsClient = createDfsClient(baseUrl, sharedKeyCredentials,
            abfsConfiguration, tokenProvider != null ? tokenProvider : null, 
            sasTokenProvider != null ? sasTokenProvider : null, encryptionContextProvider,
            abfsClientContext);
    this.blobAbfsClient = createBlobClient(baseUrl, sharedKeyCredentials,
            abfsConfiguration, tokenProvider != null ? tokenProvider : null, 
            sasTokenProvider != null ? sasTokenProvider : null, encryptionContextProvider,
            abfsClientContext);
    initServiceType(abfsConfiguration);
}

Reply:
Store is not passing null values. It expects client to have 2 constructors.
I would have to make null checks in both the constructors. To avoid redundant code I implemented createDfsClient() and createBlobCleint(). they will have both types on token providers and have null checks,